sdk/trace/span.go

```diff
@@ -225,9 +224,17 @@ func (s *span) End(options ...trace.SpanEndOption) {
 		return
 	}
 
+	// If the span was not properly started, stopwatch will be nil.
+	// The normal check (IsRecording) happens after end time generation,
+	// so we add a simple check here to make sure end time is generated
+	// as early as possible.
+	if s.stopwatch == nil {
+		return
+	}
+
 	// Store the end time as soon as possible to avoid artificially increasing
 	// the span's duration in case some operation below takes a while.
-	et := internal.MonotonicEndTime(s.startTime)
+	et := s.stopwatch.Started().Add(s.stopwatch.Elapsed())
```

Comment:
@MrAlias @Frefreak 

Do we want to "force" this implementation?
We can change the `Stopwach` implementation to:
```
type  Stopwach interface {
   Stop() time.Time
```

Pros:
- I think we could get better performance (need to be checked)
- We add full freedom on how the end time will be returned. Maybe it will be retrieved using some CGO call to retrieve it from some atomic clock?

Cons:
- Users that write some custom clock synchronization in pure Go may easily create non-monotonic time.

Reply:
My previous proposal is similar to this, but passing start time to Stopwatch instead of letting stopwatch store it. Then for default clock the stopwatch can be just a plain `struct{}`, and for an offset clock the stopwatch only needs to be a `time.Duration` to ensure the same offset is applied to both the start time and end time. In your __industrial scenario__ the version without `startTime` passed in seems reasonable.
With that being said I'm not against the `Started` and `Elapsed` implementation, after all my initial purpose was to make the current sdk has a way to use custom clock correction logic.

Reply:
> Users that write some custom clock synchronization in pure Go may easily create non-monotonic time.

I think this is the more common scenario and an important reason to not use this type of implementation. In the situation where users have access to accurate synchronized clocks it should be a CPU cycle or two to turn that into a duration which seems acceptable.

Reply:
IMO the interface doesn't need to require Started but stop watchers can implement one if they want to report their start time. Stop makes it clear that consecutive reads won't be available once the stop watcher is stopped. Elapsed give an impression that it's possible to read continuously. Stop would be useful to free the underlying resources if any, I'd +1 to Stop.

Reply:
Having the Start method means the `Span` dose not need another field to track the start time and encapsulates everything you need to determine an end time by just being passed a `Stopwatch`.

> Stop would be useful to free the underlying resources if any

That's a good point. I think having a `Stop` or `End` method could make sense, but I still think it needs to return a `time.Duration` to ensure the monotonicity.

Reply:
Sorry I missed this comment previously. From my understanding since the `Elapsed` will logically only be called once when determining end time, we should rename this `Elapsed` to `Stop` to hinted that the underlying resource (if any) can be release, right? And do you think documenting that the `Stop` will only be called once (and, return a positive `time.Duration`) a good idea?
@pellared @rakyll @MrAlias

Reply:
1.  Documenting that the Stop will only be called once is a very good idea.
2.  Documenting that it should return a positive (or non-negative) duration is a very good idea.
3. What should we do if a negative duration will be returned? Will we trim it to `0`? If so then it could be (but does not have to) documented as well.